### PR TITLE
[V1][Metrics] Support Sliding Window Attention (SWA) and hybrid layers in MFU/MBU estimation

### PR DESCRIPTION
**Pull request by thillai-c**
## Purpose

Partially addresses [#38170](https://github.com/vllm-project/vllm/issues/38170) (specifically resolving the Sliding Window Attention (SWA) and hybrid layer MFU/MBU accounting portion; Multi-Token Prediction (MTP) is tracked separately) and resolves the TODO at line 428 of `vllm/v1/metrics/perf.py`.

vLLM's analytic performance estimation module (`vllm/v1/metrics/perf.py`, enabled via `--enable-mfu-metrics`) derives Model FLOPs Utilization (MFU) and Model Bandwidth Utilization (MBU) for running models.

Previously, `AttentionMetrics` assumed full quadratic attention across all layers:
```python
TC = ctx.total_token_context_product()  # sum(num_tokens * context_len)
flops["attn_qk"] = 2 * q * TC * d * L
flops["attn_av"] = 2 * q * TC * d * L
```
For models employing **Sliding Window Attention (SWA)** (e.g., Mistral, Qwen2.5) or **hybrid/alternating layer types** (e.g., Gemma 2, Step-3.5-Flash), tokens in SWA layers only attend to at most `min(context_len, sliding_window)` tokens.

This discrepancy causes:
1. **Severe overestimation of attention core FLOPs** for long contexts (e.g., 32k or 128k context with a 4096 sliding window over-counts attention core FLOPs by up to 8×–30×, severely distorting MFU).
2. **Overestimation of decode KV-cache memory read traffic**, distorting MBU metrics.

This PR adds:
1. **`ExecutionContext` Window Bounds**:
   - Maintains per-request record tracking `(num_tokens, context_len, is_prefill)` inside `ExecutionContext`.
   - Adds window-clamped methods `total_token_context_product(sliding_window=None)`, `prefill_context_len_for_window()`, and `decode_context_len_for_window()`.
2. **`SlidingWindowAttentionParser`**:
   - Added to `AttentionMetrics.get_parser()` to parse `sliding_window` from `model_config.get_sliding_window()` or `cache_config.sliding_window`.
   - Respects `model_config.disable_sliding_window`.
   - Determines the layer breakdown (`num_swa_layers` vs `num_full_layers`) for:
     - Uniform SWA models (e.g., Mistral: all layers use sliding window).
     - Models with explicit `layer_types` (e.g., Step-3.5-Flash, Mistral-large, Hy-v4).
     - Implicit alternating architectures (e.g., Gemma 2: even layers are sliding window).
3. **`AttentionMetrics` Proportional Split**:
   - Projections (`qkv_proj`, `out_proj`) and newly generated token cache writes continue to run across all $L$ layers.
   - Core attention FLOPs (`attn_qk`, `attn_av`) and input read bytes are split proportionally between full attention layers (unwindowed) and SWA layers (window-clamped).
   - Zero regression / 100% backward-compatible for models without SWA.

## Test Plan

Added comprehensive unit tests in `tests/v1/metrics/test_perf_metrics.py` covering:
- `test_sliding_window_parser_no_swa`: Verifies standard dense models default to 0 SWA layers and `sliding_window=None`.
- `test_sliding_window_parser_uniform_swa`: Verifies full SWA models (Mistral) detect all layers as SWA.
- `test_sliding_window_parser_explicit_layer_types`: Verifies configs with explicit `layer_types` count the exact number of sliding attention layers.
- `test_sliding_window_parser_gemma2`: Verifies alternating layer detection for Gemma 2 (13 SWA out of 26 layers).
- `test_sliding_window_parser_disabled`: Verifies `disable_sliding_window=True` turns off SWA.
- `test_attention_metrics_sliding_window_flops_bounds`: Verifies that when context exceeds sliding window (8192 > 1024), core attention FLOPs are strictly bounded (8× reduction) while projection FLOPs remain unchanged.
- `test_attention_metrics_sliding_window_read_bytes`: Verifies decode KV-cache memory traffic is clamped to sliding window in SWA layers.
- `test_attention_metrics_hybrid_alternating_layers`: Verifies exact mathematical blending for models with 50% full layers and 50% SWA layers.
- `test_attention_metrics_sliding_window_per_gpu`: Verifies proper scaling across tensor parallelism and pipeline parallelism.

### Commands Run
```bash
.venv/bin/python -m pytest tests/v1/metrics/test_perf_metrics.py --confcutdir=tests/v1/metrics -v
pre-commit run ruff-check --files vllm/v1/metrics/perf.py tests/v1/metrics/test_perf_metrics.py
```

## Test Result

```text
============================= test session starts =============================
platform win32 -- Python 3.12.11, pytest-9.1.1, pluggy-1.6.0
collected 65 items

tests/v1/metrics/test_perf_metrics.py::test_base_config_parser PASSED    [  1%]
...
tests/v1/metrics/test_perf_metrics.py::test_sliding_window_parser_no_swa PASSED [ 87%]
tests/v1/metrics/test_perf_metrics.py::test_sliding_window_parser_uniform_swa PASSED [ 89%]
tests/v1/metrics/test_perf_metrics.py::test_sliding_window_parser_explicit_layer_types PASSED [ 90%]
tests/v1/metrics/test_perf_metrics.py::test_sliding_window_parser_gemma2 PASSED [ 92%]
tests/v1/metrics/test_perf_metrics.py::test_sliding_window_parser_disabled PASSED [ 93%]
tests/v1/metrics/test_perf_metrics.py::test_attention_metrics_sliding_window_flops_bounds PASSED [ 95%]
tests/v1/metrics/test_perf_metrics.py::test_attention_metrics_sliding_window_read_bytes PASSED [ 96%]
tests/v1/metrics/test_perf_metrics.py::test_attention_metrics_hybrid_alternating_layers PASSED [ 98%]
tests/v1/metrics/test_perf_metrics.py::test_attention_metrics_sliding_window_per_gpu PASSED [100%]

======================== 65 passed, 1 warning in 6.38s ========================
```

- **Linting**:
  ```text
  ruff check...............................................................Passed
  ```
- **Model Evaluation**: N/A - this change affects analytical MFU/MBU metrics accounting only; model weights, attention kernels, and output logits are unchanged.

## AI assistance

This PR was written with AI assistance (Claude Code). I reviewed every changed line, validated the behavior, and ran the tests listed above myself.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by thillai-c** [comment]
Hi @markmc, could you please take a look at this PR when you have a moment? It adds Sliding Window Attention (SWA) and hybrid layer accounting to the MFU/MBU performance metrics subsystem in v1/metrics/perf.py. Thanks!

File: vllm/v1/metrics/perf.py [hunk 1]
```diff
@@ -452,24 +571,57 @@ def get_num_flops_breakdown(
...
         T = ctx.total_num_tokens()
-        TC = ctx.total_token_context_product()
+        TC_full = ctx.total_token_context_product()
+
+        L_swa = self.num_swa_layers
+        L_full = L - L_swa
 
         if per_gpu:
             L //= self.pp_size
+            L_swa = round(L * self.num_swa_layers / self.num_hidden_layers)
```
Comment (markmc) [inline vllm/v1/metrics/perf.py]: **P1 — Hybrid SWA estimates are incorrect under pipeline parallelism.**

This preserves the total layer count, but it still rounds the SWA count *after* reducing the layer count per pipeline stage. For a 6-layer model with one SWA layer and `pp_size=2`, `L` becomes 3 and `round(3 * 1 / 6)` becomes 0. The estimate therefore treats all three layers as full attention, misreporting both attention FLOPs and `attn_input` reads.

That result is neither an actual pipeline stage’s layer mix nor the global per-GPU average. Please compute the full- and SWA-attention contributions from global layer counts and divide each by PP, or make the calculation PP-rank-aware using the actual layer partition. The regression test should cover this 6-layer / 1-SWA / PP=2 case and assert the intended semantics.

_Assisted-by: Codex (GPT-5)_

**Comment by markmc** [comment]
@thillai-c lgtm at a very high-level. I dropped one review comment in there from codex. Thanks!

File: vllm/v1/metrics/perf.py (hunk 1 above)
Comment (thillai-c) [reply vllm/v1/metrics/perf.py]: Good catch, thanks. Confirmed: with 6 layers, 1 SWA layer, and `pp_size=2`, `L` became 3 and `round(3 * 1 / 6)` evaluated to 0, so all three per-GPU layers were billed as full attention and the sliding-window contribution disappeared entirely.

Fixed by taking your first suggestion: the full-attention and SWA layer counts are now derived from the global layer counts and divided by `pp_size` independently, rather than reducing `L` first and splitting afterward. This preserves `L_full + L_swa == L` and ensures the per-GPU counts reconcile with the whole model across pipeline stages. For the case above, this now gives 2.5 full-attention layers and 0.5 SWA layers per GPU.

I didn't go with the PP-rank-aware option because these metrics are constructed by `from_vllm_config()` from config alone, with no pipeline rank available at that point. So the per-GPU number here represents the average pipeline stage rather than a specific stage. Making it rank-aware would require plumbing the rank through the parser chain, which could be handled separately if exact per-stage semantics are preferred.

One related consequence is that this also replaces the previous `L //= pp_size` flooring. That flooring silently dropped remainder layers when the model's layer count wasn't evenly divisible by the number of pipeline stages—for example, 7 layers with `pp_size=2` effectively billed only 6 layers.

This isn't really separable from the SWA fix: keeping the floored `L` while deriving the attention sublayer counts from the global counts would make the metric internally inconsistent. For 7 layers with `pp_size=2`, the attention terms would account for 3.5 layers per GPU while the projection terms would account for only 3. Using exact division keeps those components consistent.

Tests added/updated:

* `test_attention_metrics_minority_swa_survives_pipeline_parallelism` covers the exact 6-layer / 1-SWA / `pp_size=2` case, asserts the 2.5 / 0.5 split, and verifies that the result remains strictly below billing all three layers at full context.
* `test_attention_metrics_per_gpu_layers_sum_to_whole_model` covers the remainder-layer case and verifies that the per-GPU counts multiplied by `pp_size` reconcile with the whole-model counts.
* `test_attention_metrics_hybrid_sliding_window_per_gpu` previously encoded the old behavior with an 8/7 split for a 30-layer alternating model. It now expects 7.5/7.5, which represents the average per-stage split for alternating layers.

All 68 tests in `tests/v1/metrics/test_perf_metrics.py` pass. The three tests above fail against the previous implementation.

**Comment by thillai-c** [comment]
@markmc   Thanks! Pushed a fix for the codex P1: the full and SWA counts are now derived from the global layer counts and each divided by pp_size, instead of reducing L first and splitting after, which was rounding a minority SWA layer to zero. Added a regression test at the exact 6-layer / 1-SWA / pp_size=2 shape, and updated the existing hybrid PP test that had encoded the old 8/7 split. Kindly review it.

**Comment by markmc** [comment]
@thillai-c please take a look at https://docs.vllm.ai/en/latest/contributing/#ai-assisted-contributions - personally I like to see attribution of AI assistance tools :+1:

**Review by markmc — APPROVED** [review]

**Comment by thillai-c** [comment]
@markmc Thanks for the pointer. I've added an AI assistance note to the PR description. The fix commit is already part of the branch history, so I haven't rewritten its commit message. I'll include an `Assisted-by:` trailer on subsequent commits.
Claude Code assisted with this work; I reviewed every changed line and ran the tests myself.
